### DISCUSSION TITLE
high-level: response -> JSON?

### DISCUSSION THREAD

**Discussion by gutschein** [aldas/modbus-tcp-client#82] [Q&A] [answered]
I tried to get a JSON from the response just using
```
$fc3 = ReadRegistersBuilder::newReadHoldingRegisters('tcp://192.168.xxx.yyy:502',3)
->int32(4711, 'temperature [°C]')
->int32(56482, 'power [W]')
->build();
$responses = (new NonBlockingClient(['readTimeoutSec' => 0.5]))->sendRequests($fc3);
```
`l_json = json_encode($responses)`

but without success. Getting just the data part as JSON would be very helpfull. 
```
  { "data" : [ { "address" : "4711", "value" : "27", "text" : "temperature [°C]" },
               { "address" : "56482", "value" : "1234", "text" : "power [W]" } ] }
```
I can parse it in a loop. But maybe there is an easy way in the library out-of-the-box and/or PHP to solve this?
If I've the data part the rest is easy

As an Imrovement (not a need): Getting the hole thing would be perfect.
```
{ "result" : 
  { "ip" : "192.168.xxx.yyy", "port" : "502", "slaveID" : "3", "ts" : "1272509157" },
  { "data" : [ { "address" : "4711", "value" : "27", "text" : "temperature [°C]" },
               { "address" : "56482", "value" : "1234", "text" : "power [W]" } ] },
  { "errors" : [ ] }
}
```
Same for the request of the high-level
```
{ "request" : [ 
	{ "address" : "4711", "format" : "int32", "text" : "temperature [°C]"}, 
	{ "address" : "56482", "format" : "int32", "text" : "power [W]" } 
] }
```
If not via JSON: Is there any simple way to call dynamically instead of hard-coded 
```
->int32(4711, 'temperature [°C]')
->int32(65482, 'power [W]')
```
...? Or is there no way for the high-level and I've to switch to low-level?

**Comment by gutschein** [comment, marked as answer]
No answer to the JSON-question, but for "request dynamically"... I found an example ... :)
```
$fc3RequestsFromArray = ReadRegistersBuilder::newReadHoldingRegisters()
    ->allFromArray($registers)
    ->build();
```
No question about the request anymore. Just for response -> JSON

**Reply by gutschein** [reply]
Hm,,, I wonder...why it doesn't work:
```
$json = '[{"uri":"tcp://192.168.xxx.yyy:502","type":"uint32","address":"4711","name":"BatStat"},{"uri":"tcp://192.168.xxx.yyy:502","type":"int32","address":"4712","name":"BatLoadPo$
$registers = json_decode($json, true);
print_r($registers);
```
provides
```
Array
(
    [0] => Array
        (
            [uri] => tcp://192.168.xxx.yyy:502
            [type] => uint32
            [address] => 4711
            [name] => BatStat
        )

    [1] => Array
        (
            [uri] => tcp://192.168.xxx.yyy:502
            [type] => int32
            [address] => 4712
            [name] => BatLoadPower
        )

)
```
But processing 
```
$fc3 = ReadRegistersBuilder::newReadHoldingRegisters()
    ->allFromArray($registers)
    ->build();

$responses = (new NonBlockingClient(['readTimeoutSec' => 0.5]))->sendRequests($fc3);
```
provides no data, but errors. What's wrong with the array? Isn't it like "example_parallel_requests_amp.php"?

**Reply by aldas** [reply]
```$responseContainer = (new NonBlockingClient(['readTimeoutSec' => 0.2]))->sendRequests($fc3);``` 
Returns object of type [ResultContainer](https://github.com/aldas/modbus-tcp-client/blob/master/src/Network/ResultContainer.php). To get data out of it you need to use [getData()](https://github.com/aldas/modbus-tcp-client/blob/0ce52f3a7ade81f34c470249ea7dc611b5f4359f/src/Network/ResultContainer.php#L51) method on it.  This is NonBlockingClient specific. Amp and React examples return plain arrays. Amp/React does not gather allresponses to one array/object - they just echo each parsed response separately.

If you want to output parsed/extracted data as json, keyed by names, you could do  
```
echo json_encode($responseContainer->getData(), JSON_PRETTY_PRINT);
```

Example:

I will start modbus server simulator ([diagslave](https://www.modbusdriver.com/diagslave.html)) on port 5022. NB: I removed some requested addresses so only 1 request for port 5022 would be created.
```bash
./diagslave/linux_x86-64/diagslave -m tcp -p 5022
```

and run example

Output:
```bash
/usr/bin/php /home/user/code/modbus-tcp-client/examples/example_parallel_requests.php
{
    "pump2_feedbackalarm_do": false,
    "pump3_overload_alarm_do": false,
    "direction": 0,
    "battery3_voltage_wo": 0,
    "wind_angle_wo": 0,
    "gps_speed": 0,
    "distance_total_wo": 0,
    "gen2_energyw0_wo": 0,
    "longitude": 0,
    "username": ""
}
Process finished with exit code 0
```

**Reply by gutschein** [reply]
ah... in general I understand, but in OO-details I'm still an amateur... and understand afterwards...

* RESPONSE: success with your assistance (using `$responseContainer->getData()`): result as JSON. I was quite sure that there is a solution with just a tiny hint... :) There we are! That JSON was a must for me. For now my outbound requirements are solved. Thank you!

* RESPONSE: Probably more as an future improvement a JSON as I described above would be helpfull for more users with the complete picture of the results with master data, result data and errors in one JSON.

* REGISTERS: But what's still relevant for me: filling the registers dynamically via JSON, not hard-coded. I tried e.g.
```
$json = '[{"uri":"tcp://192.168.xxx.yyy:502","type":"uint32","address":"4711","name":"BatStat"},{"uri":"tcp://192.168.xxx.yyy:502","type":"int32","address":"4712","name":"BatLoadPower"}]';
$registers = json_decode($json, true); 
$fc3 = ReadRegistersBuilder::newReadHoldingRegisters()
    ->allFromArray($registers)
    ->build();
$responses = (new NonBlockingClient(['readTimeoutSec' => 0.5]))->sendRequests($fc3);
print_r($responses);
```
provides no response data, but errors. 

-> What's wrong with the `$registers`-array? Isn't it like the "example_parallel_requests_amp.php"? 
Thank you in advance for any hint how to solve a flexible register definition via JSON.

**Reply by aldas** [reply]
If you want to create register addresses for requests from array see this [ReadRegistersBuilder->fromArray()](https://github.com/aldas/modbus-tcp-client/blob/4d7b65985a18adb0f77fbda1cadd48e3a4d6eed3/src/Composer/Read/ReadRegistersBuilder.php#L85) method implementation. 

I do not see any problems with your $registers. it works

I'm using
```php
$json = '[{"uri":"tcp://127.0.0.1:5022","type":"uint32","address":"4711","name":"BatStat"},
    {"uri":"tcp://127.0.0.1:5022","type":"int32","address":"4712","name":"BatLoadPower"}]';
$registers = json_decode($json, true);
$fc3 = ReadRegistersBuilder::newReadHoldingRegisters()
    ->allFromArray($registers)
    ->build();
$responses = (new NonBlockingClient(['readTimeoutSec' => 0.5]))->sendRequests($fc3);
print_r($responses->getData());
```

or

```php
$registers = [
    ['uri' => 'tcp://127.0.0.1:5022', 'type' => 'uint32', 'address' => 4711, 'name' => 'BatStat'],
    ['uri' => 'tcp://127.0.0.1:5022', 'type' => 'int32', 'address' => 4712, 'name' => 'BatLoadPower'],
];
$fc3 = ReadRegistersBuilder::newReadHoldingRegisters()
    ->allFromArray($registers)
    ->build();

$responseContainer = (new NonBlockingClient(['readTimeoutSec' => 0.5]))->sendRequests($fc3);
print_r($responseContainer->getData());
```

Output of script:
```
/usr/bin/php /home/user/code/modbus-tcp-client/examples/example_parallel_requests_amp.php
Array
(
    [BatStat] => 0
    [BatLoadPower] => 0
)

Process finished with exit code 0
```

This is output from diagslave simulator
```bash
x@x:~/code$ ~/Downloads/diagslave/linux_x86-64/diagslave -m tcp -p 5022
diagslave 3.1 - FieldTalk(tm) Modbus(R) Diagnostic Slave Simulator
Copyright (c) 2002-2018 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: MODBUS/TCP
Slave configuration: address = -1, master activity t/o = 3.00
IP configuration: port = 5022, connection t/o = 60.00

Server started up successfully.
Listening to network (Ctrl-C to stop)

validateMasterIpAddr: accepting connection from 127.0.0.1
Slave   0: readHoldingRegisters from 4712, 3 references
..^C
```

I do not know why you are addressing 2 32bit values (uint32 and int32) as  addresses 4711 and 4712.  (u)int32 length is 2 consecutive register addresses. So value of uint32 starting at address 4711 uses address 4711 and 4712 for its 32 bits.  And int32 address at 4712 would use register 4712 and 4713 for its data.  Currently you have overlapping addresses

**Reply by aldas** [reply]
> RESPONSE: Probably more as an future improvement a JSON as I described above would be helpfull for more users with the complete picture of the results with master data, result data and errors in one JSON.

This is up to library user to implement how input gets to registers - whether it is in method of array from. And how output is dealt with is also up to implementer choosing. In this case response does not contain information about request as it is redundant and if it is need it is up to implementer to include in response. [ResultContainer](https://github.com/aldas/modbus-tcp-client/blob/master/src/Network/ResultContainer.php) API provides  `getData` and `getErrors` methods to access results for each response (in case NonBlockingClient was used).

**Reply by gutschein** [reply]
Hej Martti,
thank your for response. Good to know, that I was on the right way. I compared every char of your and my code. But I get errors, no data anyway. After a while I may found the difference: Is there a default ID = 0 and maybe your simulator uses the same ID? My devices uses slaveID = 3. 

-> How can I add the ID for using a table for the registers? Or is there a default variable like for the endian?

Your question about 4711/4712: Just sample addresses to keep the sample simple. But you're right: Makes no sense as ther are no 32bits in between... sorry for confusing. For my deviceses I use different ones. 

Your answer concerning the JSON-improvment: Understood. That's totally fine for me. Your library is great. And will be even better, if I become more familiar with it. e.g. your hint `get Errors.`

**Reply by gutschein** [reply]
Update: Looking into your code and trying I figured out...`unitId` is needed in table or json:
```
$json = '[
{"uri":"tcp://192.168.xxx.yyy:502","unitId":"3","type":"uint32","address":"4711","name":"BatStat"},
{"uri":"tcp://192.168.xxx.yyy:502","unitId":"3","type":"int32","address":"5822","name":"BatLoadPower"}
]';
```
works perfect! :)
Great work. Powerful and simple to use. Thank you!
